pkg/handlers/service_logs.go

```diff
@@ -4,88 +4,157 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
+	"regexp"
 
 	sdk "github.com/openshift-online/ocm-sdk-go"
 	slv1 "github.com/openshift-online/ocm-sdk-go/servicelogs/v1"
 	"github.com/openshift/ocm-agent-operator/api/v1alpha1"
 	"github.com/openshift/ocm-agent/pkg/consts"
-	log "github.com/sirupsen/logrus"
+	"github.com/prometheus/alertmanager/template"
 )
 
-type OCMClient interface {
-	SendServiceLog(summary, firingDesc, resolveDesc, clusterID string, severity v1alpha1.NotificationSeverity, logType string, references []v1alpha1.NotificationReferenceType, firing bool) error
+const (
+	ServiceLogActivePrefix  = "Issue Notification"
+	ServiceLogResolvePrefix = "Issue Resolution"
+)
+
+type ServiceLogBuilder struct {
+	wrappedBuilder *slv1.LogEntryBuilder
+	summary        string
+	firingDesc     string
+	resolveDesc    string
+	references     []v1alpha1.NotificationReferenceType
 }
 
-// ServiceLogsHandler manages service logs via the OCM API.
-type ServiceLogsHandler struct {
-	ocm *sdk.Connection
+type ServiceLog = slv1.LogEntry
+
+func NewServiceLogBuilder(summary, firingDesc, resolveDesc, clusterUUID string, severity v1alpha1.NotificationSeverity, logType string, references []v1alpha1.NotificationReferenceType) *ServiceLogBuilder {
+	return &ServiceLogBuilder{
+		wrappedBuilder: slv1.NewLogEntry().
+			ServiceName(consts.ServiceLogServiceName).
+			ClusterUUID(clusterUUID).
+			InternalOnly(false).
+			Severity(slv1.Severity(severity)).
+			LogType(slv1.LogType(logType)),
+		summary:     summary,
+		firingDesc:  firingDesc,
+		resolveDesc: resolveDesc,
+		references:  references,
+	}
 }
 
-// NewServiceLogsHandler creates a new handler for service logs.
-func NewServiceLogsHandler(o *sdk.Connection) *ServiceLogsHandler {
-	log.Debug("Creating new service logs handler")
-	return &ServiceLogsHandler{
-		ocm: o,
+var (
+	slVarRefRe = regexp.MustCompile(`\${[^{}]*}`)
+)
+
+func replacePlaceHoldersInString(s string, alert *template.Alert) (string, error) {
+	var err error
+	resolvePlaceHolder := func(placeHolder string) string {
+		if err == nil {
+			key, value, isOk := placeHolder[2:len(placeHolder)-1], "", false
+
+			if value, isOk = alert.Labels[key]; !isOk {
+				if value, isOk = alert.Annotations[key]; !isOk {
+					err = fmt.Errorf("alert has no '%s' label or annotation which could be used to replace place holders in the template", key)
+
+					return placeHolder
```

Comment:
Can a test be added for this block too? Also, if am reading it right, did you mean `return err` on line 60?

Reply:
Ah ok.. Got it.. :+1: